File: Demos/Delphi/Unit1.pas
unit Unit1;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls;

type

  TIBDataPumpCallBack = procedure(ARepLine: PChar); stdcall;

  TForm1 = class(TForm)
    RunSilent: TButton;
    Memo1: TMemo;
    ShowIBPump: TButton;
    procedure RunSilentClick(Sender: TObject);
    procedure ShowIBPumpClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
  private
    ABusy: boolean;
  public
    { Public declarations }
  end;

function IBPumpExec(AProfile, ASourceFile, ADestFile: PChar; ACallBack: TIBDataPumpCallBack): integer;
  stdcall; external 'IBPumpPlug.dll ';

procedure IBPumpShow;
  stdcall; external 'IBPumpPlug.dll ';

var
  Form1: TForm1;

implementation

{$R *.DFM}

procedure ShowProgress(ARepLine: PChar); stdcall;
begin
  Form1.Memo1.Lines.Add(ARepLine);
end;

procedure TForm1.FormCreate(Sender: TObject);
begin
  ABusy := False;
end;

procedure TForm1.FormCloseQuery(Sender: TObject; var CanClose: Boolean);
begin
  CanClose := not ABusy;
end;

procedure TForm1.RunSilentClick(Sender: TObject);
var
  AProfile, ASourceFile, ADestFile:string;
  Res: integer;
begin
  if not ABusy then
  begin
    ABusy := True;
    try
      Memo1.Lines.Clear;
      { IB DataPump profile }
      AProfile := 'E:\Progs\IBDataPump\MSSQL_Northwind.ibp';
      { path/connection string to source database if different from profile }
//      ASourceFile := 'd:\ibdata\bs.gdb';
      { path to dest database if different from profile }
//      ADestFile := 'd:\ibdata\bstest.gdb';
      ASourceFile := '';
      ADestFile := '';
      Res := IBPumpExec(PChar(AProfile), PChar(ASourceFile), PChar(ADestFile), ShowProgress);
      if Res = 0 then ShowMessage('All Fine.');
      if Res = -1 then ShowMessage('Error!');
      if Res > 0 then ShowMessage('Errors during pumping - ' + IntToStr(Res));
    finally
      ABusy := False;
    end;
  end;  
end;

procedure TForm1.ShowIBPumpClick(Sender: TObject);
begin
  if not ABusy then
  begin
    ABusy := True;
    try
      IBPumpShow;
    finally
      ABusy := False;
    end;
 end;
end;

end.

File: Demos/CBuilder/Unit1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

typedef void (__stdcall *LPIBDataPumpCallBack)(PChar ARepLine);

typedef int (__stdcall *LPIBPumpExec)(PChar AProfile, PChar ASourceFile, PChar ADestFile,
  LPIBDataPumpCallBack ACallBack);

typedef void (__stdcall *LPIBPumpShow)(void);

void __stdcall ShowProgress(PChar ARepLine)
{
  Form1->Memo1->Lines->Add(ARepLine);
}

//---------------------------------------------------------------------------
void __fastcall TForm1::ShowIBPumpClick(TObject *Sender)
{
  if (!m_hDLL) return;
  void* DLLFunc;
  DLLFunc = GetProcAddress(m_hDLL, "IBPumpShow");
  if (DLLFunc)
  {
    (*(LPIBPumpShow)DLLFunc)();
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::RunSilentClick(TObject *Sender)
{
  if (!m_hDLL) return;
  void* DLLFunc;
  int Res;
  AnsiString AProfile, ASourceFile, ADestFile;
  Memo1->Lines->Clear();
  AProfile = "E:\\Progs\\IBDataPump\\DbDemos1.ibp";
//  ASourceFile = "d:\\ibdata\\bs.gdb";
//  ADestFile = "d:\\ibdata\\bstest.gdb";
  ASourceFile = "";
  ADestFile = "";

  DLLFunc = GetProcAddress(m_hDLL, "IBPumpExec");
  if (DLLFunc)
  {
    Res = (*(LPIBPumpExec)DLLFunc)(AProfile.c_str(), ASourceFile.c_str(), ADestFile.c_str(), ShowProgress);
    if (Res == 0) ShowMessage("All Fine.");
    if (Res == -1) ShowMessage("Error!");
    if (Res > 0) ShowMessage("Errors during pumping - " + IntToStr(Res));
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCreate(TObject *Sender)
{
  m_hDLL = LoadLibrary("IBPumpPlug.dll");
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormDestroy(TObject *Sender)
{
  FreeLibrary(m_hDLL);
}
//---------------------------------------------------------------------------
